--- a/config/common_base
+++ b/config/common_base
@@ -626,7 +626,7 @@ CONFIG_RTE_MAX_VIRTIO_CRYPTO=32
 #
 # Compile PMD for AESNI backed device
 #
-CONFIG_RTE_LIBRTE_PMD_AESNI_MB=n
+CONFIG_RTE_LIBRTE_PMD_AESNI_MB=y
 
 #
 # Compile PMD for Software backed device
@@ -636,7 +636,7 @@ CONFIG_RTE_LIBRTE_PMD_OPENSSL=n
 #
 # Compile PMD for AESNI GCM device
 #
-CONFIG_RTE_LIBRTE_PMD_AESNI_GCM=n
+CONFIG_RTE_LIBRTE_PMD_AESNI_GCM=y
 
 #
 # Compile PMD for SNOW 3G device
