--- a/drivers/crypto/kasumi/meson.build
+++ b/drivers/crypto/kasumi/meson.build
@@ -4,9 +4,9 @@
 lib = cc.find_library('sso_kasumi', required: false)
 if not lib.found() or not cc.has_header('sso_kasumi.h')
 	build = false
-	subdir_done()
+else
+	ext_deps += lib
 endif
 
-ext_deps += lib
 sources = files('rte_kasumi_pmd.c', 'rte_kasumi_pmd_ops.c')
 deps += ['bus_vdev']
--- a/drivers/crypto/snow3g/meson.build
+++ b/drivers/crypto/snow3g/meson.build
@@ -4,10 +4,10 @@
 lib = cc.find_library('sso_snow3g', required: false)
 if not lib.found() or not cc.has_header('sso_snow3g.h')
 	build = false
-	subdir_done()
+else
+	ext_deps += lib
 endif
 
 allow_experimental_apis = true
-ext_deps += lib
 sources = files('rte_snow3g_pmd.c', 'rte_snow3g_pmd_ops.c')
 deps += ['bus_vdev', 'cryptodev']
--- a/drivers/crypto/zuc/meson.build
+++ b/drivers/crypto/zuc/meson.build
@@ -4,9 +4,9 @@
 lib = cc.find_library('sso_zuc', required: false)
 if not lib.found() or not cc.has_header('sso_zuc.h')
 	build = false
-	subdir_done()
+else
+	ext_deps += lib
 endif
 
-ext_deps += lib
 sources = files('rte_zuc_pmd.c', 'rte_zuc_pmd_ops.c')
 deps += ['bus_vdev']
