--- a/drivers/net/mlx5/linux/mlx5_os.c
+++ b/drivers/net/mlx5/linux/mlx5_os.c
@@ -2142,9 +2142,10 @@ mlx5_os_pci_probe(struct rte_pci_driver
 		dev_config.txq_inline_max = MLX5_ARG_UNSET;
 		dev_config.txq_inline_min = MLX5_ARG_UNSET;
 		dev_config.txq_inline_mpw = MLX5_ARG_UNSET;
-		dev_config.txqs_inline = MLX5_ARG_UNSET;
+		dev_config.txqs_inline = 1;
 		dev_config.vf_nl_en = 1;
 		dev_config.mr_ext_memseg_en = 1;
+		dev_config.mprq.enabled = 1;
 		dev_config.mprq.max_memcpy_len = MLX5_MPRQ_MEMCPY_DEFAULT_LEN;
 		dev_config.mprq.min_rxqs_num = MLX5_MPRQ_MIN_RXQS;
 		dev_config.dv_esw_en = 1;
